pvlib/horizon.py

```diff
@@ -0,0 +1,646 @@
+        to the origin points from which the elevation angles are to be
+        calculated. The observer points.
+
+    pt2 : ndarray
+        Nx3 array that contains lat, lon, and elev values that correspond
+        to the target points to which the elevation angles are to be
+        calculated. The observee points.
+
+    Returns
+    -------
+    bearing_deg: numeric
+        The bearings from pt1 to pt2 in degrees East of North.
+
+    elevation_angle_deg: numeric
+        The elevation angles that the points in pt2 make with the horizontal
+        as observed from the points in pt1. Given in degrees above the
+        horizontal.
+
+    '''
+    a = 6378137.0
+    b = 6356752.0
+
+    lat1 = np.atleast_1d(pt1.T[0])
+    lon1 = np.atleast_1d(pt1.T[1])
+    elev1 = np.atleast_1d(pt1.T[2])
+    lat2 = np.atleast_1d(pt2.T[0])
+    lon2 = np.atleast_1d(pt2.T[1])
+    elev2 = np.atleast_1d(pt2.T[2])
+
+    # convert to radians
+    phi1 = np.radians(lat1)
+    theta1 = np.radians(lon1)
+    phi2 = np.radians(lat2)
+    theta2 = np.radians(lon2)
+
+    v1 = tools.lle_to_xyz(np.stack([lat1, lon1, elev1], axis=1))
+    v2 = tools.lle_to_xyz(np.stack([lat2, lon2, elev2], axis=1))
+    x1 = np.atleast_1d(v1.T[0])
+    y1 = np.atleast_1d(v1.T[1])
+    z1 = np.atleast_1d(v1.T[2])
+
+    delta = np.atleast_2d(np.subtract(v1, v2))
+
+    normal = np.atleast_2d(np.stack([2*x1/a**2, 2*y1/a**2, 2*z1/b**2], axis=1))
+
+    # Take the dot product of corresponding vectors
+    dot = np.sum(np.multiply(delta, normal), axis=1)
+    beta = np.arccos(dot / np.linalg.norm(delta, axis=1)
+                     / np.linalg.norm(normal, axis=1))
+    elevation_angle = beta - np.pi/2
+
+    elevation_angle_deg = np.degrees(elevation_angle)
+
+    bearing = np.arctan2(np.sin(theta2-theta1)*np.cos(phi2),
+                         (np.cos(phi1) * np.sin(phi2)
+                          - np.sin(phi1) * np.cos(phi2)*np.cos(theta2-theta1)))
+    bearing_deg = np.degrees(bearing)
+
+    mask = (bearing_deg < 0)
+    bearing_deg[mask] += 360
+
+    return bearing_deg, elevation_angle_deg
+
+
+def calculate_horizon_points(lat_grid, lon_grid, elev_grid,
+                             sampling_method="grid", sampling_param=400):
+    """
+    Calculates a horizon profile from a three grids containing lat, lon,
+    and elevation values. The "site" is assumed to be at the center of the
+    grid.
+
+    Parameters
+    ----------
+    lat_grid : ndarray
+        A 2d array containing latitude values that correspond to the other
+        two input grids.
+
+    lon_grid : ndarray
+        A 2d array containing longitude values that correspond to the other
+        two input grids.
+
+    elev_grid : ndarray
+        A 2d array containing elevation values that correspond to the other
+        two input grids.
+
+    sampling_method : string
+        A string that specifies the sampling method used to generate the
+        horizon profile.
+
+    sampling_param : variable
```

Comment:
Suggest calling this input `num_samples`, default `None`, and describe the value that should be passed if `sampling_method` is `triangles` or `interpolator` (suggest "polar" instead of "interpolator")